### PR TITLE
Add support for ORDER BY.

### LINKED ISSUE #2096: sql: add support for ORDER BY
The initial implementation should just be a new type of `planNode` which buffers the results and sorts them on the desired columns. Follow-on optimization would be to incorporate ORDER BY requirements into index selection to eliminate the explicit sort when possible.

### PR DESCRIPTION
**Pull request by petermattis**
Sorting is currently performed in memory. Ordering is not taken into
consideration during index selection, but sorting is not performed if
the output is already in the correct order.

Fixes #2096.

### CODE REVIEW

File: sql/plan.go [hunk 1]
```diff
@@ -122,6 +122,10 @@ type planNode interface {
 	// Columns returns the column names. The length of the returned slice is
 	// guaranteed to be equal to the length of the tuple returned by Values().
 	Columns() []string
+	// The indexes of the columns the output is ordered by. Indexes are 1-based
+	// and negative indexes indicate descending ordering. The result may be nil
```
Comment (tamird) [inline sql/plan.go]: Possibly being dense: what is "the result"?

Comment (petermattis) [inline sql/plan.go]: The `[]int`. Let me reword this to make that clearer.

File: sql/scan.go [hunk 2]
```diff
@@ -261,6 +261,35 @@ func (n *scanNode) initTargets(targets parser.SelectExprs) error {
...
+		// Non-unique indexes have the primary key columns appended to their key.
+		n.columnIDs = append(n.columnIDs, n.desc.PrimaryIndex.ColumnIDs...)
+	}
+
+	// Loop over the column IDs and determine if they are used for any of the
+	// render targets.
+	n.ordering = nil
+	for _, colID := range n.columnIDs {
+		for i, r := range n.render {
+			qval, ok := r.(*qvalue)
```
Comment (tamird) [inline sql/scan.go]: nit: wouldn't this read nicer as:
```go
if qval, ok := r.(*qvalue); ok {
			if qval.col.ID == colID {
				n.ordering = append(n.ordering, i+1)
				break
			}
}
```

Comment (tamird) [inline sql/scan.go]: also can you add a comment on the type assertion? this was a cache miss for me

File: sql/select.go [hunk 3]
```diff
@@ -112,6 +121,7 @@ func (p *planner) expandSubqueries(stmt parser.Statement) error {
 func (p *planner) selectIndex(s *scanNode) (planNode, error) {
 	if s.desc == nil || s.filter == nil {
 		// No table or where-clause.
+		s.initOrdering()
```
Comment (tamird) [inline sql/select.go]: why not do this up in `(*planner).Select`?

File: sql/scan.go (hunk 2 above)
Comment (petermattis) [inline sql/scan.go]: Sure. Changed.

File: sql/sort.go [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+	// We grab a copy of columns here because we might add new render targets
+	// below. This is the set of columns requested by the query.
+	columns := s.Columns()
+	var ordering []int
+
+	for _, o := range n.OrderBy {
+		index := 0
+		switch t := o.Expr.(type) {
+		case *parser.QualifiedName:
+			if len(t.Indirect) > 0 {
```
Comment (tamird) [inline sql/sort.go]: comment on this?

File: sql/select.go (hunk 3 above)
Comment (petermattis) [inline sql/select.go]: Yeah, that's a better location. Moved.

File: sql/sort.go [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+				return nil, fmt.Errorf("qualified name \"%s\" not found", t)
```
Comment (tamird) [inline sql/sort.go]: possibly extract this error construction?

Comment (tamird) [inline sql/sort.go]: on second thought, they should probably be different errors

File: sql/select.go (hunk 3 above)
Comment (petermattis) [inline sql/select.go]: Oh, I remember now. `p.selectIndex` returns a `planNode`. Right now the `planNode` is always a `*scanNode`, but this will change fairly soon when we perform more interesting index lookups.

Comment (tamird) [inline sql/select.go]: but why not do the `initOrdering` before the `selectIndex`? i might be missing something.

File: sql/sort.go [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+				}
+			}
+			if index == 0 {
+				return nil, fmt.Errorf("qualified name \"%s\" not found", t)
+			}
+
+		default:
+			// TODO(pmattis): Use NormalizeExpr.
+			datum, err := parser.EvalExpr(o.Expr)
+			if err != nil {
```
Comment (tamird) [inline sql/sort.go]: this needs a comment

File: sql/sort.go [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+				if err := s.addRender(parser.SelectExpr{Expr: o.Expr}); err != nil {
+					return nil, err
+				}
+				index = len(s.columns)
+				break
+			}
+			i, ok := datum.(parser.DInt)
+			if !ok {
+				return nil, fmt.Errorf("invalid order by (%s)", t)
```
Comment (tamird) [inline sql/sort.go]: should the styling be `ORDER BY`?

File: sql/sort.go [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+					break
+				}
+			}
+			if index == 0 {
+				return nil, fmt.Errorf("qualified name \"%s\" not found", t)
+			}
+
+		default:
+			// TODO(pmattis): Use NormalizeExpr.
+			datum, err := parser.EvalExpr(o.Expr)
```
Comment (tamird) [inline sql/sort.go]: o.Expr is `t`, yeah?

File: sql/select.go (hunk 3 above)
Comment (petermattis) [inline sql/select.go]: The ordering depends on the index used.

File: sql/sort.go [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+func (n *sortNode) Columns() []string {
+	return n.columns
+}
+
+func (n *sortNode) Ordering() []int {
+	return n.ordering
+}
+
+func (n *sortNode) Values() parser.DTuple {
+	// If an ordering expression was used the number of columns in each row might
```
Comment (tamird) [inline sql/sort.go]: is this the case of `SELECT k from kv ORDER BY v`?

File: sql/sort.go (hunk 4 above)
Comment (petermattis) [inline sql/sort.go]: I've just reworked this because this was an incorrect limitation. The `QualifiedName` can refer to either an output column alias or to an actual column name.

File: sql/sort.go (hunk 5 above)
Comment (petermattis) [inline sql/sort.go]: After the reorg, there is only one error returned from this code path.

File: sql/sort.go (hunk 8 above)
Comment (petermattis) [inline sql/sort.go]: Yep. This has changed a bit.

File: sql/sort.go (hunk 6 above)
Comment (petermattis) [inline sql/sort.go]: Yeah, I'm assuming that `EvalExpr` returned an error because the expression contained a qualified name. Once `NormalizeExpr` is used we can return the error instead. I've added a comment to this effect.

File: sql/sort.go (hunk 7 above)
Comment (petermattis) [inline sql/sort.go]: Yep. Done.

File: sql/sort.go [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+	return len(n.values.rows)
+}
+
+func (n *sortNode) Less(i, j int) bool {
+	// TODO(pmattis): An alternative to this type of field-based comparison would
+	// be to construct a sort-key per row using encodeTableKey(). Using a
+	// sort-key approach would likely fit better with a disk-based sort.
+	ra, rb := n.values.rows[i], n.values.rows[j]
+	for _, k := range n.ordering {
+		var da, db parser.Datum
```
Comment (tamird) [inline sql/sort.go]: I'd suggest breaking this up into a comparator on the various implementations of `Datum`, rather than having all the logic in this method here.

File: sql/sort.go (hunk 9 above)
Comment (petermattis) [inline sql/sort.go]: Yes. Or `SELECT k, v FROM kv ORDER BY k+v`.

File: sql/sort.go (hunk 10 above)
Comment (petermattis) [inline sql/sort.go]: Ok. I've added `Datum.Compare` and provided implementations. This did dramatically simplify this method, pushing the guts into the datum implementations.

File: sql/parser/eval.go [hunk 11]
```diff
@@ -65,6 +68,21 @@ func (d DBool) Type() string {
 	return "bool"
 }
 
+// Compare implements the Datum interface.
+func (d DBool) Compare(other Datum) int {
+	v, ok := other.(DBool)
```
Comment (tamird) [inline sql/parser/eval.go]: i'm not sure this matters now, but this choice deserves mention

Comment (petermattis) [inline sql/parser/eval.go]: Which choice? That anything other than a `DBool` compares greater.

File: sql/sort.go [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+					}
+				}
+			}
+
+			if index == 0 {
+				if err := qname.NormalizeColumnName(); err != nil {
+					return nil, err
+				}
+				if qname.Table() == "" || equalName(s.desc.Alias, qname.Table()) {
+					for j, r := range s.render {
```
Comment (tamird) [inline sql/sort.go]: i'm missing something here; how come this doesn't need to iterate over the table's columns?

File: sql/parser/eval.go (hunk 11 above)
Comment (tamird) [inline sql/parser/eval.go]: yeah, and that the same is true for all `Datum`s, meaning sort order is only partially defined

Comment (tamird) [inline sql/parser/eval.go]: (except null)

Comment (petermattis) [inline sql/parser/eval.go]: I've added explanatory comments to the implementations. The expectation is that all the values in a column will either have a particular datum type or `dNull`. We shouldn't ever have to compare a `DInt` vs a `DString`. But now that I write that I'm not sure if it is true as expressions like `CASE` can return different types. Let me leave it as-is for now with a `TODO` in the `sortNode` code.

File: sql/sort.go (hunk 12 above)
Comment (petermattis) [inline sql/sort.go]: I should add some more comments. The above loop over `s.columns` is checking to see if the qualified name matches one of the output columns. If we can't find a match, this loop is checking to see if the qualified name matches one of the existing render targets. This might also fail, in which case we add a new render target to the scanNode down on line 79.

**Comment by petermattis** [comment]
Rebased to pick up `NormalizeExpr` and added some more comments to `planner.orderBy`.

**Comment by petermattis** [comment]
Any other comments here?

**Comment by tamird** [comment]
looking

File: sql/parser/eval.go [hunk 13]
```diff
@@ -113,6 +180,32 @@ func (d DTuple) Type() string {
 	return "tuple"
 }
 
+// Compare implements the Datum interface.
+func (d DTuple) Compare(other Datum) int {
+	v, ok := other.(DTuple)
+	if !ok {
+		// Anything other than a DTuple compares greater (e.g. `(1, 2) > NULL`).
```
Comment (tamird) [inline sql/parser/eval.go]: giving `NULL` as the example here is just a bit misleading. I understand that's the case you're solving for, though.

File: sql/sort.go [hunk 14]
```diff
@@ -0,0 +1,224 @@
...
+								break
+							}
+						}
+					}
+				}
+			}
+		}
+
+		if index == 0 {
+			// The order by expression matched neither an output column or an
```
Comment (tamird) [inline sql/sort.go]: s/ or / nor /

File: sql/sort.go [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+			// existing render target.
+			if datum, ok := expr.(parser.Datum); ok {
+				// If we evaluated to an int, use that as an index to order by.
```
Comment (tamird) [inline sql/sort.go]: can you give an example here?

File: sql/sort.go [hunk 16]
```diff
@@ -0,0 +1,224 @@
...
+	}
+	return n.values.Next()
+}
+
+func (n *sortNode) Err() error {
+	return n.err
+}
+
+// wrap the supplied planNode with the sortNode if sorting is required.
+func (n *sortNode) wrap(plan planNode) planNode {
```
Comment (tamird) [inline sql/sort.go]: could you add a TODO that mentions reverse scans? i'm not sure where that would go.

File: sql/sort.go (hunk 14 above)
Comment (petermattis) [inline sql/sort.go]: Done.

**Comment by tamird** [comment]
LGTM. Since you're making some final additions, mind squashing this?

File: sql/sort.go (hunk 15 above)
Comment (petermattis) [inline sql/sort.go]: Examples added here and elsewhere in this function.

File: sql/sort.go (hunk 16 above)
Comment (petermattis) [inline sql/sort.go]: Reverse scan support will go along with using ordering for index selection. I've started work on that but probably won't handle reverse scans initially. I'll add a TODO to that code (which isn't in this PR).

**Comment by petermattis** [comment]
Squashed.
